pydantic_ai_graph/pydantic_ai_graph/graph.py

```diff
@@ -2,169 +2,232 @@
 
 import base64
 import inspect
-import types
-from dataclasses import dataclass
-from datetime import datetime, timezone
+from dataclasses import dataclass, field
 from pathlib import Path
 from time import perf_counter
 from typing import TYPE_CHECKING, Any, Generic
 
 import logfire_api
-from typing_extensions import TypeVar, assert_never
+from typing_extensions import Never, ParamSpec, Protocol, TypeVar, assert_never
 
 from . import _utils
-from .nodes import BaseNode, End, GraphContext, GraphOutputT, NodeDef
-from .state import Snapshot, StateT
+from ._utils import get_parent_namespace
+from .nodes import BaseNode, End, GraphContext, NodeDef
+from .state import EndEvent, GraphHistoryItem, StateT, Step
 
-__all__ = ('Graph',)
+__all__ = ('Graph', 'GraphRun', 'GraphRunner')
 
 _logfire = logfire_api.Logfire(otel_scope='pydantic-ai-graph')
-GraphInputT = TypeVar('GraphInputT', default=Any)
+
+RunSignatureT = ParamSpec('RunSignatureT')
+RunEndT = TypeVar('RunEndT', default=None)
+NodeRunEndT = TypeVar('NodeRunEndT', covariant=True, default=Never)
+
+
+class StartNodeProtocol(Protocol[RunSignatureT, StateT, NodeRunEndT]):
+    def get_id(self) -> str: ...
+    def __call__(self, *args: RunSignatureT.args, **kwargs: RunSignatureT.kwargs) -> BaseNode[StateT, NodeRunEndT]: ...
 
 
-# noinspection PyTypeHints
 @dataclass(init=False)
-class Graph(Generic[StateT, GraphInputT, GraphOutputT]):
+class Graph(Generic[StateT, RunEndT]):
     """Definition of a graph."""
 
-    first_node: NodeDef[StateT, Any, Any]
-    nodes: dict[str, NodeDef[StateT, Any, Any]]
-    name: str | None
+    name: str
+    nodes: tuple[type[BaseNode[StateT, RunEndT]], ...]
+    node_defs: dict[str, NodeDef[StateT, RunEndT]]
 
     def __init__(
         self,
-        first_node: type[BaseNode[StateT, GraphInputT, GraphOutputT]],
-        *other_nodes: type[BaseNode[StateT, Any, GraphOutputT]],
-        name: str | None = None,
+        name: str = 'graph',
+        nodes: tuple[type[BaseNode[StateT, RunEndT]], ...] = (),
```

Comment:
I would prefer `*nodes: BaseNode[StateT, RunEndT]]`

Reply:
There were a couple reasons I changed this:
* I liked having the name before the nodes (though I didn't realize the plan was to infer it from the assignment statement like with agent, which reduces that issue).
* I like being able to see where the top-level items start/end more clearly. Using an actual tuple here ensures they are indented relative to the other top-level kwargs in `Graph`. Without that, you get a bunch of items listed and then top-level kwargs mixed in.

	Concretely, I just find it visually easier to parse:
	```python
	g = Graph(
	    nodes=(
	        Float2String,
	        String2Length,
	        Double,
	    ),
	    state_type=MyState,
	)
	```
	vs.
	```python
	g = Graph(
	    Float2String,
	    String2Length,
	    Double,
	    state_type=MyState,
	)
	```
	Don't you? (And this only gets worse as you get more node types and more kwargs.)
	
	But I can revert this change if you insist.